### DISCUSSION TITLE
Add "header" config to input-autocomplete-api interface

### DISCUSSION THREAD

**Discussion by loongmxbt** [directus/directus#12750] [Feature Requests]
Many third party apis requires a header with Authorization tokens, I think adding a header json config to `input-autocomplete-api interface` would be useful.

```
headers:{  
        "Authorization": xxx-token
    }  
```

Following is some third-party api with header.
```
var options = {  
    hostname:'open.api.xxx.com',
    method:'GET',  
    path:`/services/open/search/2.0?word=${value}&pageSize=20&pageNum=1`,  
    headers:{  
        "Authorization": token
    }  
}  
```

**Comment by azrikahar** [comment]
Should be the same as #11192 👍

**Reply by loongmxbt** [reply]
Yes, found the same issue, is someone working on this?
We're not sure whether to make a custom interface simply adding a json config or modify the default `input-autocomplete-api interface` and make a PR.

**Reply by azrikahar** [reply]
Personally I'm not aware of anyone working on this. Custom interface is definitely a workable escape hatch as always, but perhaps the existing autocomplete api interface can be enhanced with the same "use repeater to add headers" approach as #8855 did for webhooks. But I'll defer to @rijkvanzanten for his thoughts on this enhancement 😄

**Reply by rijkvanzanten** [reply]
Makes sense to me! PRs for this are welcome 🙂

**Reply by loongmxbt** [reply]
Our team discussed this feature and changed our thoughts, we may not need "auth headers" in the default `input-autocomplete-api interface` for security reasons.
If we directly use `open.api.xxx.com/search` with auth header configuration, anyone uses Directus Admin can see the token from browser console. It would be easy to leak. (If some api needs a token, it usually needs to pay.)
We should move the token into .env file and make it an internal endpoint to prevent it from outside use.
And #11620 means we can use a customize interface to fill in other fields resulted from our third party api.

**Reply by azrikahar** [reply]
That's a good point! Perhaps another alternative is to use a custom endpoint extension to act as a proxy, where the auth token is configured in it, so that the input-autocomplete-api can just target `your.directus.com/your-custom-endpoint` _without_ needing the api token instead. You can then lock down that custom endpoint by checking the accountability as well. Granted I have yet to actually try this to make sure it will 100% work, but sounds workable on paper 🤔

**Reply by loongmxbt** [reply]
Here are some simplified code using a custom endpoint with `input-autocomplete-api interface`:
```js
const axios = require('axios').default;

export default (router, { services, exceptions, env }) => {
  const { ItemsService } = services;
  const { ServiceUnavailableException } = exceptions;
  router.get('/companySearch', async (req, res, next) => {
    let { name } = req.query
    try {
      const { data } = await axios.get('http://open.api.tyc.com/services/open/search/2.0', { 
        params: {
          word: name,
          pageSize: 20,
          pageNum: 1
        },
        headers: {
          'Authorization': env.CUSTOM_TOKEN
        }
      })
      res.json(data)
    } catch (error) {
      return next(new ServiceUnavailableException(error.message));
    }
  })
}
```

```
.env
CUSTOM_TOKEN="xxxxxx"
```

<img width="615" alt="image" src="https://user-images.githubusercontent.com/5986600/167974960-3641b7e7-ae80-45e6-b6b1-78396f28e29a.png">

**Comment by rfos** [comment]
I'm getting 403 Forbidden when I try to retrieve values from other collectiosn using search.
Logged in as Administrator.

Request URL: http://localhost:8055/items/model?search=L80
Request Method: GET
Status Code: 403 Forbidden
Remote Address: 127.0.0.1:8055
Referrer Policy: strict-origin-when-cross-origin

If I change the visibility to public in the collection named "models" I can get all values as result.

![2022-04-18 22_21_04-Editing Item in Teste](https://user-images.githubusercontent.com/217889/163873268-9c8b26fd-3d2d-4f8b-b97b-8f6a325b5ff9.png)

**Reply by rijkvanzanten** [reply]
This makes sense. The interface makes a public network call as any outside request would, and doesn't include the authentication tokens. What are you trying to do? You might be better of with the many to one interface if you're looking to select an item from a different collection 👍🏻

**Reply by rfos** [reply]
Yes! I was trying to retrieve a value from a search in API results. I changed it to M2O and it works, but in the related table I have 800 records and it can be a little annoying to find the right one from a list. 
Thank you for you suggestions I'm going ahead in M2O alternative

**Comment by hermesalvesbr** [comment]
Is there a way to put the "bearer token" in external api? In autocomplete api field?

**Reply by rijkvanzanten** [reply]
Not as of right now. This is the open feature request for that 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
